enhancement showcase (ajax, gravatar, nested, delete)

codeshell opened the pull request:
Hi there,

added some functionality for my own use. Since some of it is listed in #22 you might be able to reuse some of it, namely:

- ajax submission
- gravatar support
- styling
- nested comments
- delete comments as admin,super (from frontend)

Two things to be aware of:
- I don't care about the backend presentation (admin panel) so that's most likely broken.
- I changed the save logic to use the current page folder which is my personal preference. Comments are stored with the markdown file they belong to, Mainly because of three reasons:
  1. changing routes doesn't orphan comments
  2. deleting pages physically deletes the corresponding comments
  3. comments are git-synced

Should be seen as WIP as it is _not_ heavily tested.

Regards,
Thorsten

**Screenshot 1:**
![image](https://user-images.githubusercontent.com/23580812/31979224-d682e42a-b945-11e7-98cb-81900883a77d.png)

**Screenshot 2:**
![image](https://user-images.githubusercontent.com/23580812/31979175-8b9f8d28-b945-11e7-8199-e6d172165cef.png)

nickbroon commented:
This appears to also help with #51 by moving the comments to the page folder.

voythas commented:
This is epic. Thank you @codeshell for this, I was looking for something like this for a long time and I'm not skilled enough to write my own.

codeshell commented:
Thanks for your kind words. Anybody experimenting with the code feel free to notify me of any issues you run into. Meanwhile I will update this PR from time to time as I see fit for my current use case.

It might help as an alternative to #60 and seems to tackle #51 indeed. I really like the suggestions #48 and #49 by Salvatos so I added variants of them to the backend with the recent commit to be used in twig templates as needed.

foufrix commented:
super nice ! thanks for that ! i'm gonna take a look right away.

Why no merge from the author ?

codeshell commented:
Shot in the dark: Maybe the theme you are using does not include FontAwesome by default? I use FontAwesome v4 to render all the symbols within the preconfigured templates and css.
Make sure, `<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" type="text/css" rel="stylesheet" />` is included before `<link href="/dev/github/grav/user/plugins/comments/assets/comments.css" type="text/css" rel="stylesheet" />`.

alorszut commented:
Hi all,

I haven't seen any commit including a good admin for comments, based ont codeshell enhancements to the backend. I just added my code modification to following files to get something in the backend Admin :

 \user\plugins\comments\admin\templates\comments.html.twig
change line 92 :
`                        <td class="details"><strong>Page</strong>: <a href="{{ base_url_simple }}/{{comment.pageTitleURL}}" target="_blank">{{comment.pageTitle}}</a><br>
`
\user\plugins\comments\comments.php :
line 718 & 719
`        	$data['comments'][$i]['pageTitle'] = $data['comments'][$i]['title'];
		$data['comments'][$i]['pageTitleURL'] = $data['comments'][$i]['lang'].$data['comments'][$i]['page'];
`
line 814 to 833

        $pages = [];
		$title_page=[];
		$count_title_page=[];
        foreach($files as $file) {
			for ($i=0; $i <count($file->data['comments']);$i++)
			{
				$title_page[$i]= $file->data['comments'][$i]['title'];
				$pages[$file->data['comments'][$i]['title']] = [
					'title' => $file->data['comments'][$i]['title'],
					'commentsCount' => count($file->data['comments']),
					'lastCommentDate' => $file->data['comments'][$i]['date']
				];
			}		
        }
		$count_title_page=array_count_values ($title_page);
		
		foreach ($count_title_page as $key => $value) {
			$pages[$key]['commentsCount'] = $value;
		}
`

Hopefully it helps.
Everybody is looking for something to delete comments from Admin, and I was guessing why not doing the same thing as done for GUESTBOOK plugin with a moderation capability and deletion/approval from admin ?

Cheers,
Alorszut

leetNightshade commented:
Nice changes! I was working on my own branch when I saw your changes. So I merged them, and am still testing them out. I started to update your stuff to look more like Reddit comments, and it's looking promising. I'm still working on changes, and evaluating. But it looks good so far. Commenting currently disabled for testing, which is something I added in my own (would be cool to add support per post), so I haven't updated add/delete visuals and ajax version. 

![image](https://user-images.githubusercontent.com/40641737/42785560-d936d946-8907-11e8-9002-a84eab45b8fc.png)

abooks23 commented:
Hi 
please attach edited files or edited all plugin(plugin it is better)

leetNightshade commented:
> 
> 
> Hi
> please attach edited files or edited all plugin(plugin it is better)

If you're talking to me, abook23, my files are [here](http://git.leetnightshade.com/leetNightshade/grav-plugin-comments).

abooks23 commented:
thanks!

abooks23 commented:
Hi
how to count page comments?
{{page.commentsCount}} not working

codeshell commented:
if you use leetNightshades code, `{{page.commentsCount}}` should work when derived from `grav.twig.pages  `( demo template: http://git.leetnightshade.com/leetNightshade/grav-plugin-comments/src/branch/develop/admin/templates/comments.html.twig ) 

stats in my PR are different and can be used with `{{stats.global_stats.active_entries}}`, `{{stats.global_stats.active_comments}}`, etc when derived from `stats = recent_comments(limit|default(5), pages_limit|default(3))` (demo template: https://github.com/getgrav/grav-plugin-comments/blob/bd1799cc7e377d9770fab3e1ba2ca71b9209679b/templates/partials/recentcomments.html.twig)

codeshell commented:
> Big thing I'm not a fan of is switch to YAML for storing comments. I don't think it adds anything, and I'm trying to make sure my site does as little processing as possible on my server. YAML is useful for human authored files, in this case comments are all processed with script, there's no reason not to pick the fastest most convenient serialization format; which is not YAML. JSON works well.
> 
> [Edit] I've been trying to figure out how to get new field in comments.yaml to show up for entering website, but I haven't been able to get it to work. Any ideas? Starting to dig through docs to see if I find out any reason why it's not showing up.

You were right that YAML is a difficult choice. However I don't think, I changed it from JSON to YAML. It has always been YAML.

What I did was changing from `File` Facade to CompiledYamlFile. 
```php
use Grav\Common\File\CompiledJsonFile;
use Grav\Common\File\CompiledYamlFile;
use RocketTheme\Toolbox\File\File;
```
That means encoding and decoding/parsing is done behind the scenes. Just replace `CompiledYamlFile` with `CompiledJsonFile` and everything will be treated and saved as JSON.

As for the locations: Have a look at https://github.com/getgrav/grav-plugin-comments/pull/65/commits/e8417c6a0ec0a6b6a95e6155d9b8f2ffcf99ee60
The original behaviour was that all comments were stored within DATA_DIR /comments/$lang/$path.yaml That code is still active (see line 265) to not break the admin panel. I think that location is a bad choice (for reasons pointed out in my original post) and decided to save comments with the page instead (see line 188). And I added a global index file containing all comments (see line 228). That one is for administrative and backup reasons. No comment is ever deleted from that index file. Just marked as deleted (soft delete). In case you delete pages and loose comments you didn't mean to delete.

To add additional fields just look at `public function onFormProcessed(Event $event)` and `public function saveComment($route, $path, $parent_id, $lang, $text, $name, $email, $title)`. 

Performance shouldn't be too much of an issue for the per-page-comment-files. Index files on the other hand are convenience and can be ditched.

abooks23 commented:
> {{stats.global_stats.active_comments}}

Hi
yes its work but it render ALL COMMENTS in all post, how render comments for current post?

abooks23 commented:
why
![Screenshot_2019-06-13 My new Blog post Grav](https://user-images.githubusercontent.com/27723595/59412852-0b5a8200-8dc7-11e9-8941-59eddf2e1077.png)
after reload ok

leetNightshade commented:
Ajax inline reply doesn't work, idk if I broke that or if it was already like that. It's something I've had to fix. Sorry I didn't mention it. It's why my websites comments are displayed, and commenting is disabled. I haven't found the time to fix it.   o.o;

abooks23 commented:
thanks

codeshell commented:
Comments and Ajax are still working for me. Using my adjustments with grav v1.6.10.

rhukster commented:
BTW, this is very cool, but because it changes some default stuff, including the location of comments, I can't merge as folks will lose their comments.

I suggest you fork, and release your own version?
